src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -174,12 +219,12 @@ public void Dispose()
                     {
                         // Since we put all of our work in a queue, we know that if we had tried to advise file or directory changes,
                         // it must have happened before now
-                        if (_directoryFilePath != null)
+                        foreach (var cookie in _directoryWatchCookies)
                         {
-                            await service.UnadviseDirChangeAsync(_directoryWatchCookie).ConfigureAwait(false);
+                            await service.UnadviseDirChangeAsync(cookie).ConfigureAwait(false);
                         }
 
-                        // it runs after disposed. so no lock is needed for _activeFileWatchingTokens
+                        // Since this runs after disposed. so no lock is needed for _activeFileWatchingTokens
```

Comment:
1. boy is that a terrifying comment.  i'm going to presuem there's a big caveat of "because if someone else was accessing this concurrently, they'd be violating the disposal contract, and then all bets are off.
2. would IAsyncDisposable be preferable here?

Reply:
1. Yes, this is the standard "after dispose, it doesn't matter" case.
2. I don't think so, since we wouldn't want it being awaited -- if I close a project there's no reason to block that action on our ability to stop watching files.

Reply:
ok.  worth doc'ing that.

Reply:
Comments updated.